### PR TITLE
testdriver-vendor.js could use a cleanup + type annotations

### PR DESCRIPTION
**Pull request by marcoscaceres**
#### b1c0630822f90f6aeb042a2010dd3f0adeb3031e
<pre>
testdriver-vendor.js could use a cleanup + type annotations
<a href="https://bugs.webkit.org/show_bug.cgi?id=248379">https://bugs.webkit.org/show_bug.cgi?id=248379</a>

Reviewed by Tim Nguyen.

Adds TypeScript type annotations to testdriver-vendor.js and cleans up the code a bit:

 * Let the JS engine handle throwing for any missing methods/properties.
 * Geolocation now only emits a single location, and only when permission is &quot;granted&quot;.

Previously, the latter was causing compounding position objects to be emitted, which would confuse various tests.

* LayoutTests/imported/w3c/web-platform-tests/resources/testdriver-vendor.js:
(logDebug):
(async dispatchMouseActions):
(window.test_driver_internal.action_sequence):
(async switch):
(window.test_driver_internal.set_permission):
(async context):
(window.test_driver_internal.minimize_window):
(async if):
(window.test_driver_internal.set_window_rect):
(await.new.Promise): Deleted.

Canonical link: <a href="https://commits.webkit.org/257315@main">https://commits.webkit.org/257315@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/6b900e389c569b32a81c2d8ec7e9a75655ce625e

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/98584 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/7794 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/31708 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/108006 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/168270 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/8304 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/85172 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/91118 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/105985 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/104266 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/6292 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/89847 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/33284 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/88110 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/21206 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/76218 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/1727 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/22735 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/1640 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/45232 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5038 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/6578 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/42166 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/3027 "Built successfully") | | |

### CODE REVIEW

File: LayoutTests/imported/w3c/web-platform-tests/resources/testdriver-vendor.js [hunk 1]
```diff
@@ -289,89 +308,97 @@ window.test_driver_internal.action_sequence = async function(sources)
...
+window.test_driver_internal.set_permission = async function(permission_params)
 {
     switch (permission_params.descriptor.name) {
     case "geolocation":
-        if (!window.testRunner?.setGeolocationPermission)
-            throw new Error("unimplemented");
-        setInterval(() => {
-            window.testRunner.setMockGeolocationPosition(51.478, -0.166, 100);
-        }, 100);
+        await pause(100);
```
Comment (nt1m) [inline LayoutTests/imported/w3c/web-platform-tests/resources/testdriver-vendor.js]: setInterval != setTimeout :)

**Review by nt1m — CHANGES_REQUESTED** [review]
Looks fine, aside from the removal of the error handling, many WPT can be ran by just starting a http server, and opening the raw file in the browser (FF/WebKit/Chrome), and this is where the error handling comes in useful.

File: LayoutTests/imported/w3c/web-platform-tests/resources/testdriver-vendor.js (hunk 1 above)
Comment (marcoscaceres) [reply LayoutTests/imported/w3c/web-platform-tests/resources/testdriver-vendor.js]: ah yeah... seemed weird that it kept setting it to the same position but I guess it's just reporting a value. 

Reminds me... we should maybe add something Web Driver for testing Geolocation.

**Comment by marcoscaceres** [comment]
> Looks fine, aside from the removal of the error handling, many WPT can be ran by just starting a http server, and opening the raw file in the browser (FF/WebKit/Chrome), and this is where the error handling comes in useful.

I agree, but the checks are redundant as the browser will throw when trying to use or call our internal objects if they are not there. IMO, it's better to just let JS throw "naturally" when calling stuff that's missing.

Does that make sense or am I missing something?

**Comment by nt1m** [comment]
> > Looks fine, aside from the removal of the error handling, many WPT can be ran by just starting a http server, and opening the raw file in the browser (FF/WebKit/Chrome), and this is where the error handling comes in useful.
> 
> I agree, but the checks are redundant as the browser will throw when trying to use or call our internal objects if they are not there. IMO, it's better to just let JS throw "naturally" when calling stuff that's missing.
> 
> Does that make sense or am I missing something?

I'd rather throw explicit errors, than get `Unhandled promise rejection: ...`. It's also pretty common practice across the codebase fwiw.

**Comment by marcoscaceres** [comment]
>I'd rather throw explicit errors, than get Unhandled promise rejection: .... It's also pretty common practice across the codebase fwiw

The "unhandled promise rejections" are going to happen irrespective though. The difference here is that you would get, for example, `"TypeError: x.foo is not a function. (In 'x.foo()', 'x.foo' is undefined)"`, which is much more helpful than "Not implemented", which doesn't provide very useful information as it doesn't actually tell you *what* is not implemented.

This solves for that.

**Review by nt1m — APPROVED** [review]

**Pull request merged**